# 1.tokenizer,构造输入

- tokenizer, model :相匹配，tokenizer outputs  => model input
- Auto*Tokenizer,Auto*Model: Generic type,Auto开头的会自适应的找到对应的model和Tokenizer,服务于model input
    - len(input_ids) == len(attention_mask)
    - tokenizer(test_senteces[0],) ==> tokenizer.\_\_call\_\_ ==>encode
    - tokenizer.encode 约等于 tokenizer.tokenize次元 + tokenizer.convert_tokens_to_ids id；缺少起止标识符101，102
    - tokenizer.decode 
    - tokenizer 的工作原理主要依赖于tokenizer.vocab字典存储了tokens ==> ids 的映射关系
        - tokenizer.special_tokens_map 记录了一些特殊符号 ==> tokens
    - attenmask = 1 的部分为句子原始部分，attemask = 0 的部分指的时padding部分 
    - truncation 与 max_length 配合一起作用

In [1]:
test_senteces = ['today is not that bad', 'today is so bad']
model_name = 'distilbert-base-uncased-finetuned-sst-2-english'

In [3]:
from transformers import AutoTokenizer, AutoModelForSequenceClassification

In [28]:
import torch

In [4]:
tokenizer = AutoTokenizer.from_pretrained(model_name)
model = AutoModelForSequenceClassification.from_pretrained(model_name)

config.json:   0%|          | 0.00/629 [00:00<?, ?B/s]

C:\Users\juju\AppData\Roaming\Python\Python312\site-packages\huggingface_hub\file_download.py:130: UserWarning: `huggingface_hub` cache-system uses symlinks by default to efficiently store duplicated files but your machine does not support them in C:\Users\juju\.cache\huggingface\hub\models--distilbert-base-uncased-finetuned-sst-2-english. Caching files will still work but in a degraded version that might require more space on your disk. This warning can be disabled by setting the `HF_HUB_DISABLE_SYMLINKS_WARNING` environment variable. For more details, see https://huggingface.co/docs/huggingface_hub/how-to-cache#limitations.
To support symlinks on Windows, you either need to activate Developer Mode or to run Python as an administrator. In order to activate developer mode, see this article: https://docs.microsoft.com/en-us/windows/apps/get-started/enable-your-device-for-development
  warnings.warn(message)


tokenizer_config.json:   0%|          | 0.00/48.0 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/268M [00:00<?, ?B/s]

Loading weights:   0%|          | 0/104 [00:00<?, ?it/s]

In [ ]:
batch_input = tokenizer(test_senteces, truncation = True,  padding=True,return_tensors='pt')


In [9]:
batch_input

{'input_ids': tensor([[ 101, 2651, 2003, 2025, 2008, 2919,  102],
        [ 101, 2651, 2003, 2061, 2919,  102,    0]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1],
        [1, 1, 1, 1, 1, 1, 0]])}

In [12]:
tokenizer.tokenize(test_senteces[0])

['today', 'is', 'not', 'that', 'bad']

In [15]:
tokenizer.convert_tokens_to_ids(tokenizer.tokenize(test_senteces[0]))

[2651, 2003, 2025, 2008, 2919]

In [13]:
tokenizer.encode(test_senteces[0])

[101, 2651, 2003, 2025, 2008, 2919, 102]

In [11]:
tokenizer(test_senteces[0],)

{'input_ids': [101, 2651, 2003, 2025, 2008, 2919, 102], 'token_type_ids': [0, 0, 0, 0, 0, 0, 0], 'attention_mask': [1, 1, 1, 1, 1, 1, 1]}

In [16]:
tokenizer.decode([101, 2651, 2003, 2025, 2008, 2919, 102])

'[CLS] today is not that bad [SEP]'

In [ ]:
tokenizer.vocab#token:id
#尾盘匹配就是unkonwn

{'[unused805]': 810,
 'skylar': 27970,
 'revealing': 8669,
 'tractor': 16358,
 '##ways': 14035,
 'eliza': 13234,
 'gift': 5592,
 'ᄑ': 1468,
 '##kov': 7724,
 'gardner': 11764,
 'rewarded': 14610,
 'practical': 6742,
 'shelter': 7713,
 'doorway': 7086,
 'definitive': 15764,
 'coefficients': 21374,
 'supervise': 28589,
 '1830s': 20400,
 '[unused839]': 844,
 '##town': 4665,
 'buena': 27493,
 'diver': 17856,
 'mean': 2812,
 'sculptors': 28417,
 'magnolia': 24659,
 '##nosis': 27109,
 'beside': 3875,
 '1702': 26776,
 'installations': 14111,
 '##eral': 21673,
 'psalm': 22728,
 '##ever': 22507,
 '##uring': 12228,
 've': 2310,
 '##gel': 12439,
 'plume': 26888,
 'absolute': 7619,
 'church': 2277,
 'director': 2472,
 'was': 2001,
 'sired': 26940,
 'chuckled': 10252,
 'hired': 5086,
 '##ে': 29917,
 'croatia': 8097,
 '##uni': 19496,
 'resorts': 16511,
 'daylight': 11695,
 'behaviour': 9164,
 'apostolic': 11815,
 '31': 2861,
 '##aka': 11905,
 'abnormal': 19470,
 '##秀': 30454,
 '##lke': 28143,
 'carro

In [20]:
tokenizer.vocab_size

30522

In [21]:
tokenizer.special_tokens_map

{'unk_token': '[UNK]',
 'sep_token': '[SEP]',
 'pad_token': '[PAD]',
 'cls_token': '[CLS]',
 'mask_token': '[MASK]'}

In [22]:
tokenizer.convert_tokens_to_ids(x for x in tokenizer.special_tokens_map.values())

[100, 102, 0, 101, 103]

---

## 📖 AI补充说明：代码中出现的核心概念详解

### 1. **token_type_ids**（片段类型标识符）
**作用**：区分输入中的不同句子/片段，主要用于**句子对任务**（Sentence Pair Tasks）

**工作原理**：
- `0`：表示第一个句子/片段
- `1`：表示第二个句子/片段

**应用场景**：
- **问答系统**：区分问题(0)和答案(1)
- **自然语言推理**：区分前提(0)和假设(1)  
- **句子相似度**：区分两个待比较的句子

**本例中的情况**：
```python
'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0],  # 全为0，因为是单句分类任务
                         [0, 0, 0, 0, 0, 0, 0]])
```
由于情感分析是**单句分类任务**，所以所有token的type都是0。这个字段在单句任务中可以忽略，但模型仍会返回它。

---

### 2. **attention_mask**（注意力掩码）
**作用**：告诉模型哪些token是真实的，哪些是padding填充的

**数值含义**：
- `1`：真实token，需要模型关注
- `0`：padding token，应该被忽略

**示例**：
```python
'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1],  # 第一句无padding
                         [1, 1, 1, 1, 1, 1, 0]]) # 第二句末尾有padding(0)
```

---

### 3. **truncation 与 padding 的组合使用**

| 参数组合 | 效果 | 适用场景 |
|---------|------|---------|
| `padding=True` | 填充到batch内最长句子长度 | 句子长度差异不大 |
| `padding='max_length'` | 填充到模型最大长度(通常512) | 需要固定长度输入 |
| `truncation=True` | 超过最大长度时截断 | 处理长文本 |
| `max_length=32` | 自定义最大长度 | 特定长度要求 |

**推荐做法**：
```python
# 方案1：动态长度（节省内存）
tokenizer(texts, padding=True, truncation=True, return_tensors='pt')

# 方案2：固定长度（适合批处理）
tokenizer(texts, max_length=128, padding='max_length', 
          truncation=True, return_tensors='pt')
```

---

### 4. **模型输出的三个关键属性**

```python
SequenceClassifierOutput(
    loss=None,              # 训练时才有，推理时为None
    logits=tensor(...),     # 未归一化的预测分数
    hidden_states=None,     # 隐藏层状态（需配置输出）
    attentions=None         # 注意力权重（需配置输出）
)
```

**logits**：
- 经过神经网络最后一层但未经过softmax的原始输出
- 值越大表示模型越倾向于该类别
- 需要通过softmax转换为概率

---

### 5. **ID2Label 映射**
```python
"id2label": {
    "0": "NEGATIVE",
    "1": "POSITIVE"
}
```
- 将模型输出的数字ID转换为可读的标签
- `label2id` 则是反向映射

---

### 6. **特殊Token的ID对照表**

| Token | 含义 | ID | 用途 |
|-------|------|-----|------|
| `[UNK]` | Unknown | 100 | 词汇表中不存在的词 |
| `[SEP]` | Separator | 102 | 句子分隔符 |
| `[PAD]` | Padding | 0 | 填充符 |
| `[CLS]` | Classification | 101 | 句子起始符（用于分类） |
| `[MASK]` | Mask | 103 | 掩码符（用于预训练） |

---

> **💡 提示**：以上内容由AI根据Hugging Face Transformers库最新文档补充，帮助理解2022年视频未涵盖的新版本特性。



In [ ]:
tokenizer(test_senteces, truncation=True, padding=True, return_tensors='pt')#指定最大长度，超长截断，补齐, padding = true 于max_length不兼容，当指定padding = True时，遍历所有句子，以所有句子的最长长度为padding的尺寸

In [24]:
tokenizer(test_senteces, max_length=32, truncation=True,
          padding='max_length', return_tensors='pt')  # 指定最大长度，超长截断，补齐

{'input_ids': tensor([[ 101, 2651, 2003, 2025, 2008, 2919,  102,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0],
        [ 101, 2651, 2003, 2061, 2919,  102,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,    0,
            0,    0,    0,    0,    0,    0,    0,    0]]), 'token_type_ids': tensor([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0],
        [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0]]), 'attention_mask': tensor([[1, 1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0],
        [1, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
         0, 0, 0, 0, 0, 0, 0, 0]])}

# 2.mode，调用模型 & parse output ，输出解析

In [31]:
import torch.nn.functional as F

In [35]:
model.config

DistilBertConfig {
  "activation": "gelu",
  "architectures": [
    "DistilBertForSequenceClassification"
  ],
  "attention_dropout": 0.1,
  "bos_token_id": null,
  "dim": 768,
  "dropout": 0.1,
  "dtype": "float32",
  "eos_token_id": null,
  "finetuning_task": "sst-2",
  "hidden_dim": 3072,
  "id2label": {
    "0": "NEGATIVE",
    "1": "POSITIVE"
  },
  "initializer_range": 0.02,
  "label2id": {
    "NEGATIVE": 0,
    "POSITIVE": 1
  },
  "max_position_embeddings": 512,
  "model_type": "distilbert",
  "n_heads": 12,
  "n_layers": 6,
  "output_past": true,
  "pad_token_id": 0,
  "qa_dropout": 0.1,
  "seq_classif_dropout": 0.2,
  "sinusoidal_pos_embds": false,
  "tie_weights_": true,
  "tie_word_embeddings": true,
  "transformers_version": "5.1.0",
  "vocab_size": 30522
}

In [38]:
with torch.no_grad():
    outputs = model(**batch_input)
    print(outputs)
    score = F.softmax(outputs.logits, dim=1)
    print(score)
    labels = torch.argmax(score,dim=1)
    print(labels)
    labels = [model.config.id2label[id] for id in labels.tolist()]
    print(labels)

SequenceClassifierOutput(loss=None, logits=tensor([[-3.4620,  3.6118],
        [ 4.7508, -3.7899]]), hidden_states=None, attentions=None)
tensor([[8.4631e-04, 9.9915e-01],
        [9.9980e-01, 1.9531e-04]])
tensor([1, 0])
['POSITIVE', 'NEGATIVE']


经过一系列神经网络，送到softmax 之前的结果一般来说叫做logits

默认模型
In [1]: from transformers import pipeline
In [2]: task_name = 'sentiment-analysis'

In [3]: pipeline(task_name)
No model was supplied, defaulted to distilbert/distilbert-base-uncased-finetuned-sst-2-english and revision 714eb0f.
Using a pipeline without specifying a model name and revision in production is not recommended.

---

## 🚀 Transformers库最新版本（v5.x）的重要变化

### 与2022年版本的主要差异

#### 1. **默认模型变更**
```python
# 旧版本（2022）
task_name = 'sentiment-analysis'
pipeline(task_name)  
# 默认: distilbert-base-uncased-ffinetuned-sst-2-english

# 新版本（2024+）
pipeline(task_name)
# 默认: distilbert/distilbert-base-uncased-ffinetuned-sst-2-english
# 注意：模型路径增加了组织名称前缀
```

#### 2. **safe_tensors 格式成为标准**
```python
# 旧版本：模型文件为 .bin (PyTorch格式)
# model.bin

# 新版本：默认使用 .safetensors (更安全、更快的加载格式)
# model.safetensors
```

**优势**：
- ✅ 安全性：防止恶意模型文件
- ✅ 速度：加载速度更快
- ✅ 跨平台：更好的兼容性

#### 3. **新推荐的最佳实践**

**A. 使用Auto Classes时指定模型修订版本**：
```python
# 推荐：指定revision以确保可复现性
model = AutoModelForSequenceClassification.from_pretrained(
    "distilbert/distilbert-base-uncased-finetuned-sst-2-english",
    revision="714eb0f"  # 固定特定版本
)

# 不推荐：每次可能加载不同版本
model = AutoModelForSequenceClassification.from_pretrained(
    "distilbert/distilbert-base-uncased-finetuned-sst-2-english"
)
```

**B. 设置环境变量避免警告**：
```python
import os
os.environ['HF_HUB_DISABLE_SYMLINKS_WARNING'] = '1'
```

**C. 使用tokenizers的更快实现**：
```python
# 新版本默认使用Rust实现的tokenizer（速度快10-100倍）
tokenizer = AutoTokenizer.from_pretrained(model_name)
# 无需额外配置，自动使用fast tokenizer
```

#### 4. **输出对象的新方法**

```python
outputs = model(**batch_input)

# 新增方法：直接获取预测结果
predictions = outputs.logits.argmax(dim=-1)

# 新增：使用后处理函数更方便
from transformers import pipeline

classifier = pipeline("sentiment-analysis", model=model_name)
results = classifier(test_senteces)
# 输出：[{'label': 'POSITIVE', 'score': 0.9991}, 
#        {'label': 'NEGATIVE', 'score': 0.9998}]
```

---

### 代码现代化建议

#### ❌ 旧写法
```python
import torch
with torch.no_grad():  # 手动管理梯度
    outputs = model(**inputs)
    logits = outputs.logits
    predictions = torch.argmax(logits, dim=1)
```

#### ✅ 新写法（推荐）
```python
# 方法1：使用pipeline（最简单）
from transformers import pipeline
classifier = pipeline("sentiment-analysis", model=model_name)
results = classifier(test_senteces)

# 方法2：使用trainer（训练时）
from transformers import Trainer
trainer = Trainer(model=model)
predictions = trainer.predict(test_dataset)

# 方法3：使用inference模式（推理时）
model.eval()  # 设置为评估模式
with torch.inference_mode():  # 比no_grad()更快
    outputs = model(**inputs)
```

---

### 性能优化提示

```python
# 1. 启用缓存（加速重复输入）
tokenizer = AutoTokenizer.from_pretrained(model_name)

# 2. 批处理（比单个处理快得多）
texts = ["text1", "text2", "text3", ...]  # 一次处理多个
inputs = tokenizer(texts, padding=True, return_tensors='pt')

# 3. 使用半精度（在支持GPU时）
model = AutoModelForSequenceClassification.from_pretrained(
    model_name, 
    torch_dtype=torch.float16  # 使用FP16加速
)
```

---

> **📝 版本信息**：
> - 视频录制时间：2022年
> - 当前Transformers版本：5.1.0+
> - 主要变化：模型命名规范、safetensors格式、性能优化



---

## 📚 Transformers Pipeline 标准任务名称大全

### 什么是 Pipeline 任务名称？

**Pipeline 任务名称**是 Hugging Face Transformers 库中定义的**标准化任务标识符**。使用这些标准名称，`pipeline()` 函数可以自动：
- 匹配对应的 Pipeline 类
- 加载匹配的 Tokenizer、Model、Processor
- 提供统一的 API 接口

**示例**：
```python
from transformers import pipeline

# 方式1：仅指定任务名（自动选择默认模型）
classifier = pipeline("sentiment-analysis")
# 等价于：pipeline(model="distilbert/distilbert-base-uncased-finetuned-sst-2-english")

# 方式2：指定任务和模型
classifier = pipeline("sentiment-analysis", model="nlptown/bert-base-multilingual-uncased-sentiment")

# 方式3：仅指定模型（自动推断任务）
classifier = pipeline(model="distilbert/distilbert-base-uncased-finetuned-sst-2-english")
```

---

### 📝 自然语言处理任务

| 任务名称 | 别名/描述 | 默认模型示例 |
|---------|----------|------------|
| **`sentiment-analysis`** | 文本分类（特指情感分析） | `distilbert/distilbert-base-uncased-finetuned-sst-2-english` |
| **`text-classification`** | 通用文本分类 | 同上 |
| **`text-generation`** | 文本生成 | `openai-community/gpt2` |
| **`ner`** | 命名实体识别 | `dbmdz/bert-large-cased-finetuned-conll03-english` |
| **`question-answering`** | 抽取式问答 | `deepset/roberta-base-squad2` |
| **`fill-mask`** | 掩码填充 | `google-bert/bert-base-uncased` |
| **`summarization`** | 文本摘要 | `facebook/bart-large-cnn` |
| **`translation`** | 机器翻译 | `facebook/mbart-large-50-many-to-many-mmt` |
| **`text2text-generation`** | 文本到文本生成 | `google/flan-t5-base` |
| **`zero-shot-classification`** | 零样本分类 | `facebook/bart-large-mnli` |
| **`table-question-answering`** | 表格问答 | `google/tapas-base-finetuned-wtq` |

**使用示例**：
```python
# 情感分析
classifier = pipeline("sentiment-analysis")
result = classifier("I love this product!")
# 输出: [{'label': 'POSITIVE', 'score': 0.9998}]

# 命名实体识别
ner = pipeline("ner", aggregation_strategy="simple")
result = ner("My name is Wolfgang and I live in Berlin")
# 输出: [{'entity_group': 'PER', 'word': 'Wolfgang'}, 
#        {'entity_group': 'LOC', 'word': 'Berlin'}]

# 问答
qa = pipeline("question-answering")
result = qa(question="Where do I live?", context="My name is Wolfgang and I live in Berlin")
# 输出: {'score': 0.91, 'start': 34, 'end': 40, 'answer': 'Berlin'}
```

---

### 🔊 音频任务

| 任务名称 | 别名/描述 | 默认模型示例 |
|---------|----------|------------|
| **`audio-classification`** | 音频分类 | `superb/wav2vec2-base-superb-ks` |
| **`automatic-speech-recognition`** | 语音识别 | `openai/whisper-base` |
| **`text-to-speech`** | 文本转语音 | `suno/bark-small` |
| **`text-to-audio`** | 文本转音频（同上） | `suno/bark-small` |
| **`zero-shot-audio-classification`** | 零样本音频分类 | `laion/clap-htsat-unfused` |

**使用示例**：
```python
# 语音识别
asr = pipeline("automatic-speech-recognition")
result = asr("https://huggingface.co/datasets/Narsil/asr_dummy/resolve/main/1.flac")

# 音频分类
audio_classifier = pipeline("audio-classification")
result = audio_classifier("https://example.com/audio.wav")
# 输出: [{'score': 0.997, 'label': 'cat'}, ...]
```

---

### 🖼️ 计算机视觉任务

| 任务名称 | 别名/描述 | 默认模型示例 |
|---------|----------|------------|
| **`image-classification`** | 图像分类 | `microsoft/beit-base-patch16-224-pt22k-ft22k` |
| **`object-detection`** | 目标检测 | `facebook/detr-resnet-50` |
| **`image-segmentation`** | 图像分割 | `facebook/detr-resnet-50-panoptic` |
| **`depth-estimation`** | 深度估计 | `LiheYoung/depth-anything-base-hf` |
| **`video-classification`** | 视频分类 | `microsoft/swin-tiny-patch4-window7-224` |
| **`zero-shot-image-classification`** | 零样本图像分类 | `google/siglip-so400m-patch14-384` |
| **`zero-shot-object-detection`** | 零样本目标检测 | `google/owlvit-base-patch32` |
| **`image-to-image`** | 图像到图像转换 | `caidas/swin2SR-classical-sr-x2-64` |
| **`mask-generation`** | 掩码生成 | `facebook/sam-vit-base` |

**使用示例**：
```python
# 图像分类
classifier = pipeline("image-classification")
result = classifier("https://example.com/image.jpg")
# 输出: [{'score': 0.44, 'label': 'macaw'}, ...]

# 目标检测
detector = pipeline("object-detection")
result = detector("https://example.com/image.jpg")
# 输出: [{'score': 0.99, 'label': 'cat', 'box': {'xmin': 69, 'ymin': 171, ...}}]

# 零样本图像分类
classifier = pipeline("zero-shot-image-classification")
result = classifier(
    "https://example.com/image.jpg",
    candidate_labels=["cat", "dog", "bird"]
)
```

---

### 🔀 多模态任务

| 任务名称 | 别名/描述 | 默认模型示例 |
|---------|----------|------------|
| **`visual-question-answering`** / **`vqa`** | 视觉问答 | `dandelin/vilt-b32-finetuned-vqa` |
| **`document-question-answering`** | 文档问答 | `impira/layoutlm-document-qa` |
| **`image-text-to-text`** | 图像文本到文本 | `Salesforce/blip-image-captioning-base` |
| **`any-to-any`** | 通用多模态生成 | `google/gemma-3-27b-it` |

**使用示例**：
```python
# 视觉问答
vqa = pipeline("visual-question-answering")
result = vqa(
    image="https://example.com/image.jpg",
    question="What is in this image?"
)
# 输出: [{'score': 0.94, 'answer': 'a cat'}]

# 图像描述生成
captioner = pipeline("image-to-text", model="Salesforce/blip-image-captioning-base")
result = captioner("https://example.com/image.jpg", text="A photo of")
# 输出: [{'generated_text': 'a photo of two birds'}]
```

---

### ⚙️ 特征提取

| 任务名称 | 描述 | 适用模型 |
|---------|------|---------|
| **`feature-extraction`** | 文本特征提取 | 所有文本模型 |
| **`image-feature-extraction`** | 图像特征提取 | 所有视觉模型 |

**使用示例**：
```python
# 提取文本特征
extractor = pipeline("feature-extraction", model="google-bert/bert-base-uncased")
features = extractor("This is a test")
# 输出形状: torch.Size([1, 8, 768]) - 1个样本, 8个token, 768维特征

# 提取图像特征
img_extractor = pipeline("image-feature-extraction", model="google/vit-base-patch16-224")
features = img_extractor("https://example.com/image.jpg")
# 输出形状: torch.Size([1, 197, 768])
```

---

### 💡 实用技巧

#### 1. **查看所有支持的任务**
```python
from transformers import pipeline, PIPELINE_REGISTRY

# 查看所有支持的任务
for task in sorted(PIPELINE_REGISTRY.get_supported_tasks()):
    print(task)
```

#### 2. **查看模型支持的任务**
```python
from transformers import AutoConfig

config = AutoConfig.from_pretrained("distilbert/distilbert-base-uncased-finetuned-sst-2-english")
print(config.id2label)  # 查看标签映射
```

#### 3. **生产环境推荐写法**
```python
import os

# 推荐：明确指定模型、版本和token
classifier = pipeline(
    task="sentiment-analysis",
    model="distilbert/distilbert-base-uncased-finetuned-sst-2-english",
    revision="714eb0f",  # 固定版本，确保可复现
    token=os.getenv("HF_TOKEN")  # 提高下载速率
)

# 设置环境变量避免警告
os.environ['HF_HUB_DISABLE_SYMLINKS_WARNING'] = '1'
```

#### 4. **批量处理**
```python
# 批量处理比单个处理快得多
texts = ["I love this!", "This is bad", "Amazing product!"]
classifier = pipeline("sentiment-analysis")
results = classifier(texts)
# 一次处理多个，自动利用批处理加速
```

---

### 🎯 任务选择指南

**如果你的任务是...** → **使用这个任务名称**

- 判断文本情感（正面/负面） → `"sentiment-analysis"`
- 识别文本中的人名、地名、组织名 → `"ner"`
- 回答基于给定段落的问题 → `"question-answering"`
- 生成文章摘要 → `"summarization"`
- 翻译文本 → `"translation_xx_to_yy"`
- 识别图片中的物体 → `"object-detection"`
- 将图片转换为文字描述 → `"image-to-text"`
- 语音转文字 → `"automatic-speech-recognition"`
- 文字转语音 → `"text-to-speech"`

---

> **📚 参考资料**：
> - [Hugging Face Transformers Pipeline 官方文档](https://huggingface.co/docs/transformers/main_classes/pipelines)
> - [Hugging Face Hub 任务标签](https://huggingface.co/docs/hub/models-tasks)
> - [Pipeline 教程](https://huggingface.co/docs/transformers/pipeline_tutorial)

